// mcp_msg/ClientTermHandle.js
var CronJob = require("cron").CronJob;
var async = require('async');
var moment = require("moment");
var dc = require('mcp_db').dc;
var prop = require('mcp_config').prop;
var esut = require("easy_util");
var dateUtil = esut.dateUtil;

var log = esut.log;
var digestUtil = esut.digestUtil;
var termSer = require("mcp_service").termSer;

var path = require("path");
var mcpUtil = require("mcp_util");
var fsUtil = mcpUtil.fsUtil;

var cons = require('mcp_constants');
var termStatus = cons.termStatus;
var msgStatus = cons.msgStatus;
var msgType = cons.msgType;

var issue = require("mcp_issue");
var issueModel = issue.issueModel;

var ClientTermHandle = function(){};

ClientTermHandle.prototype.handle = function(msg, dTerm, cb)
{
    var self = this;
    async.waterfall([
        //处理消息
        function(cb)
        {
            if(dTerm.status == termStatus.PRE_ON_SALE)
            {
                self.open(msg, dTerm, function(err, data){
                    cb(err);
                });
            }
            else if(dTerm.status == termStatus.PREEND)
            {
                self.close(msg, dTerm, function(err, data){
                    cb(err);
                });
            }
            else if(dTerm.status == termStatus.PRE_SEND)
            {
                self.backClose(msg, dTerm, function(err, data){
                    cb(err);
                });
            }
            else if(dTerm.status == termStatus.IN_CALCULATE)
            {
                self.check(msg, dTerm, function(err, data){
                    cb(err);
                });
            }
            else if(dTerm.status == termStatus.IN_PAYOUT)
            {
                self.prize(msg, dTerm, function(err, data){
                    cb(err);
                });
            }
            else
            {
                cb(null);
            }
        },
        //标记消息已经处理完成
        function(cb)
        {
            var table = dc.mg_msg.get("msg");
            table.findAndModify({_id:msg._id}, {},
            {$set:{status:msgStatus.HANDLED}}, [], function(err, data){
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, result);
    });
}

/**
 * 期次开售
 * @param msg
 * @param detail_term
 * @param cb
 */
ClientTermHandle.prototype.open = function(msg, dTerm, cb)
{
    log.info(dTerm);
    cb(null);
}

/**
 * 期次停售
 * @param msg
 * @param detail_term
 * @param cb
 */
ClientTermHandle.prototype.close = function(msg, dTerm, cb)
{
    log.info(dTerm);
    cb(null);
}


/**
 * 期次后台停售
 * @param msg
 * @param detail_term
 * @param cb
 */
ClientTermHandle.prototype.backClose = function(msg, dTerm, cb)
{
    issueModel.backClose(msg, dTerm, cb);
}

/**
 * 算奖
 * @param msg
 * @param detail_term
 * @param cb
 */
ClientTermHandle.prototype.check = function(msg, dTerm, cb)
{
    log.error("算奖开始时间"+new Date().getTime());
    var currDate = dateUtil.toCurrentDate();
    var dir = "/data/app/issue";
    dir = path.join(dir,dTerm.gameCode,currDate,dTerm.code);
    //创建文件目录
    fsUtil.createDirSync(dir);
    issueModel.check(msg, dTerm,dir, cb);
}


/**
 * 返奖
 * @param msg
 * @param detail_term
 * @param cb
 */
ClientTermHandle.prototype.prize = function(msg, dTerm, cb)
{
    issueModel.prize(msg, dTerm, cb);
}

module.exports = new ClientTermHandle();
