// mcp_issue/IssueModel.js
var CronJob = require("cron").CronJob;
var async = require('async');
var moment = require("moment");
var dc = require('mcp_db').dc;
var prop = require('mcp_config').prop;
var esut = require("easy_util");
var log = esut.log;
var digestUtil = esut.digestUtil;
var service = require("mcp_service");
var termSer = service.termSer;
var msgSer = service.msgSer;
var ticketSer = service.ticketSer;
var notifySer = service.notifySer;
var termReportSer = service.termReportSer;

var cons = require('mcp_constants');
var ticketStatus = cons.ticketStatus;
var termStatus = cons.termStatus;
var msgStatus = cons.msgStatus;
var msgType = cons.msgType;
var termReportType = cons.termReportType;

var cp = require('mcp_cp');
var checkFac = cp.checkFac;

var IssueModel = function(){};

/**
 * 查找需要处理的期次
 * @param cb
 */
IssueModel.prototype.findToHandle = function(cb)
{
    var self = this;
    async.waterfall([
        //是否需要有开售的期次
        function(cb)
        {
            termSer.findToOpen(function(err, term){
                if(term)
                {
                    msgSer.saveTerm(term, function(err, data){
                        cb(err);
                    });
                }
                else
                {
                    cb(null);
                }
            });
        },
        //是否有需要停售的期次
        function(cb)
        {
            termSer.findToClose(function(err, term){
                if(term)
                {
                    msgSer.saveTerm(term, function(err, data){
                        if(err)
                        {
                            cb(err);
                        }
                        else
                        {
                            notifySer.saveTerm(term, function(err, data){
                                cb(err);
                            });
                        }
                    });
                }
                else
                {
                    cb(null);
                }
            });
        },
        //是否有需要后台停售的期次
        function(cb)
        {
            termSer.findToBackClose(function(err, term){
                if(term)
                {
                    msgSer.saveTerm(term, function(err, data){
                        cb(err);
                    });
                }
                else
                {
                    cb(null);
                }
            });
        },
        //是否有需要算奖的期次
        function(cb)
        {
            termSer.findToCheck(function(err, term){
                if(term)
                {
                    msgSer.saveTerm(term, function(err, data){
                        if(err)
                        {
                            cb(err);
                        }
                        else
                        {
                            notifySer.saveTerm(term, function(err, data){
                                cb(err);
                            });
                        }
                    });
                }
                else
                {
                    cb(null);
                }
            });
        },
        //是否有需要返奖的期次
        function(cb)
        {
            termSer.findToPrize(function(err, term){
                if(term)
                {
                    msgSer.saveTerm(term, function(err, data){
                        cb(err);
                    });
                }
                else
                {
                    cb(null);
                }
            });
        }
    ], function (err, result) {
        cb(err, result);
    });
}

/**
 * 后台停售
 * @param msg
 * @param dTerm
 * @param cb
 */
IssueModel.prototype.backClose = function(msg, dTerm, cb)
{
    var self = this;
    var table = dc.mg.get("printqueen");
    var cond = {gameCode:dTerm.gameCode, termCode:dTerm.code};
    var hasNext = true;
    async.whilst(
        function () { return hasNext;},
        function (callback) {
            table.findAndRemove(cond, {}, {}, function(err, ticket){
                if(ticket)
                {
                    ticket.id = ticket._id;
                    ticketSer.printFail(ticket, dTerm.code, function(err, data){
                        callback(err);
                    });
                }
                else
                {
                    hasNext = false;
                    callback();
                }
            });
        },
        function (err) {
            cb(err, null);
        }
    );
}

/**
 * 算奖
 * @param cb
 */
IssueModel.prototype.check = function(msg, dTerm, cb)
{
    var self = this;
    var colName = dTerm.gameCode + "_" + dTerm.code + "_" + msg.sep;
    var conn = dc.mg.getConn();
    var table = conn.collection(colName);
    var cursor = table.find({}, {}, {});
    async.waterfall([
        //获得奖级
        function(cb)
        {
            var cond = {gameCode:dTerm.gameCode, termCode:dTerm.code};
            var sort = {level:1};
            var gradeTable = dc.main.get("gamegrade");
            gradeTable.find(cond, {level:1, bonus:1}).sort(sort).toArray(function(err, data){
                log.info(err);
                log.info(data);
                if(err)
                {
                    cb(err);
                    return;
                }
                var check = checkFac.getCheck(dTerm.gameCode, data, dTerm.wNum);
                cb(null, check);
            });
        },
        //查看一共有多少条记录
        function(check, cb)
        {
            cursor.count(function(err, count){
                cursor.close();
                cb(err, check, count);
            });
        },
        //遍历每一条记录
        function(check, count, cb)
        {
            log.info(colName + ",开始遍历记录.......");
            var curCount = 0;
            var hasNext = true;
            async.whilst(
                function () { return hasNext;},
                function (callback) {
                    table.findAndRemove({}, {}, [], function(err, ticket){
                        if(err)
                        {
                            callback(err);
                            return;
                        }
                        if(ticket)
                        {
                            curCount++;
                            var name = 'count' + ticket.pType + ticket.bType;
                            var bonusInfo = check[name]({number:ticket.number});
                            ticket.dNumber = dTerm.wNum;
                            ticket.auditTermCode = dTerm.code;
                            //保存中獎或者未中獎通知
                            self.sendMsg(ticket, bonusInfo, function(err, data){
                                if(err)
                                {
                                    callback(err);
                                    return;
                                }
                                //更新數據庫
                                ticketSer.updateBonusInfo(ticket, function(err, data){
                                    callback(err);
                                })
                            })
                        }
                        else
                        {
                            hasNext = false;
                            callback();
                        }
                    });
                },
                function (err) {
                    cb(err, count, curCount);
                }
            );
        }
    ], function (err, total, count) {
        log.info(err);
        log.info(colName + ",算奖结束,总记录数:" + total + ",完成数:" + count + ".");
        if(err)
        {
            cb(err, {total:total, finished:count});
        }
        else
        {
            if(total > 0)
            {
                conn.dropCollection(colName, function(err, data){
                    log.info("删除完成的临时记录表:" + colName);
                    log.info(err);
                    cb(err, {total:total, finished:count});
                });
            }
            else
            {
                cb(err, {total:total, finished:count});
            }
        }
    });
}

/**
 * 对单张票进行算奖
 * @param msg
 * @param dTerm
 * @param cb
 */
IssueModel.prototype.checkTicket = function(ticket)
{
    var self = this;
}


IssueModel.prototype.sendMsg = function(ticket, bonusInfo, cb)
{
    var self = this;
    ticket.id = ticket._id;
    ticket.bonus = bonusInfo.bonus;
    ticket.bonusBeforeTax = bonusInfo.bonusBeforeTax;
    ticket.bonusDetail = bonusInfo.bonusDetail;
    if(ticket.bonus > 0)
    {
        ticket.status = ticketStatus.HIT;
    }
    else
    {
        ticket.status = ticketStatus.NOT_HIT;
    }
    delete ticket._id;
    delete ticket.version;
    delete ticket.takeTime;
    delete ticket.printId;
    notifySer.saveTicket(ticket, function(err, data){
        cb(err, data);
    });
}


/**
 * 返奖
 * @param cb
 */
IssueModel.prototype.prize = function(msg, dTerm, cb)
{
    var self = this;
    var conn = dc.main.getConn();
    var now = new Date().getTime();
    var rst = {};
    var cond = "gameCode='" + dTerm.gameCode + "' and auditTermCode='" + dTerm.code + "'";
    async.waterfall([
        //销售角度统计中奖数目
        function(cb)
        {
            var sql = "select customerId,count(*) as hitCount,sum(amount) as hitAmount,sum(bonus) as bonus,sum(bonusBeforeTax) as bonusBeforeTax from tticket where status=1200 and " + cond + " group by customerId";
            conn.execute(sql, {}, function(err, data){
                for(var key in data)
                {
                    var set = data[key];
                    var key = set.customerId + "_" + termReportType.SALE;
                    var obj;
                    if(rst[key] == undefined)
                    {
                        obj = {type:termReportType.SALE, createTime:now};
                        rst[key] = obj;
                    }
                    else
                    {
                        obj = rst[key];
                    }
                    obj.customerId = set.customerId;
                    obj.hitCount = set.hitCount;
                    obj.hitAmount = set.hitAmount;
                    obj.bonus = set.bonus;
                    obj.bonusBeforeTax = set.bonusBeforeTax;
                }
                cb(err);
            });
        },
        //销售角度统计未中奖数目
        function(cb)
        {
            var sql = "select customerId,count(*) as notHitCount,sum(amount) as notHitAmount from tticket where status=1300 and " + cond + " group by customerId";
            conn.execute(sql, {}, function(err, data){
                for(var key in data)
                {
                    var set = data[key];
                    var key = set.customerId + "_" + termReportType.SALE;
                    var obj;
                    if(rst[key] == undefined)
                    {
                        obj = {type:termReportType.SALE, createTime:now};
                        rst[key] = obj;
                    }
                    else
                    {
                        obj = rst[key];
                    }
                    obj.customerId = set.customerId;
                    obj.notHitCount = set.notHitCount;
                    obj.notHitAmount = set.notHitAmount;
                }
                cb(err);
            });
        },
        //销售统计失败数目
        function(cb)
        {
            var sql = "select customerId,count(*) as failCount,sum(amount) as failAmount from tticket where status=1400 and " + cond + " group by customerId";
            conn.execute(sql, {}, function(err, data){
                for(var key in data)
                {
                    var set = data[key];
                    var key = set.customerId + "_" + termReportType.SALE;
                    var obj;
                    if(rst[key] == undefined)
                    {
                        obj = {type:termReportType.SALE, createTime:now};
                        rst[key] = obj;
                    }
                    else
                    {
                        obj = rst[key];
                    }
                    obj.customerId = set.customerId;
                    obj.failCount = set.failCount;
                    obj.failAmount = set.failAmount;
                }
                cb(err);
            });
        },
        //出票角度统计中奖数目
        function(cb)
        {
            var sql = "select printId as customerId,count(*) as hitCount,sum(amount) as hitAmount,sum(bonus) as bonus,sum(bonusBeforeTax) as bonusBeforeTax from tticket where status=1200 and " + cond + " group by printId";
            conn.execute(sql, {}, function(err, data){
                for(var key in data)
                {
                    var set = data[key];
                    var key = set.customerId + "_" + termReportType.PRINT;
                    var obj;
                    if(rst[key] == undefined)
                    {
                        obj = {type:termReportType.PRINT, createTime:now};
                        rst[key] = obj;
                    }
                    else
                    {
                        obj = rst[key];
                    }
                    obj.customerId = set.customerId;
                    obj.hitCount = set.hitCount;
                    obj.hitAmount = set.hitAmount;
                    obj.bonus = set.bonus;
                    obj.bonusBeforeTax = set.bonusBeforeTax;
                }
                cb(err);
            });
        },
        //出票角度统计未中奖数目
        function(cb)
        {
            var sql = "select printId as customerId,count(*) as notHitCount,sum(amount) as notHitAmount from tticket where status=1300 and " + cond + " group by printId";
            conn.execute(sql, {}, function(err, data){
                for(var key in data)
                {
                    var set = data[key];
                    var key = set.customerId + "_" + termReportType.PRINT;
                    var obj;
                    if(rst[key] == undefined)
                    {
                        obj = {type:termReportType.PRINT, createTime:now};
                        rst[key] = obj;
                    }
                    else
                    {
                        obj = rst[key];
                    }
                    obj.customerId = set.customerId;
                    obj.notHitCount = set.notHitCount;
                    obj.notHitAmount = set.notHitAmount;
                }
                cb(err);
            });
        },
        //出票统计失败数目
        function(cb)
        {
            var sql = "select printId as customerId,count(*) as failCount,sum(amount) as failAmount from tticket where status=1400 and " + cond + " group by printId";
            conn.execute(sql, {}, function(err, data){
                for(var key in data)
                {
                    var set = data[key];
                    var key = set.customerId + "_" + termReportType.PRINT;
                    var obj;
                    if(rst[key] == undefined)
                    {
                        obj = {type:termReportType.PRINT, createTime:now};
                        rst[key] = obj;
                    }
                    else
                    {
                        obj = rst[key];
                    }
                    obj.customerId = set.customerId;
                    obj.failCount = set.failCount;
                    obj.failAmount = set.failAmount;
                }
                cb(err);
            });
        }
    ], function (err, count) {
        async.waterfall([
            //保存报表信息
            function(cb)
            {
                termReportSer.saveList(dTerm, rst, function(err, data){
                    cb(err, data);
                });
            },
            //销售端进行返奖，出票端先支付出票的款，再扣奖金
            function(rst, cb)
            {
                async.eachSeries(rst, function(rpt, callback) {
                    termReportSer.handleMoney(rpt, function(err, data){
                        callback(err);
                    });
                }, function(err){
                    cb(err, null);
                });
            }
        ], function (err, data) {
            cb(err, data);
        });
    });
}

module.exports = new IssueModel();
