// mcp_control/AdminControl.js
var async = require('async');
var esut = require('easy_util');
var log = esut.log;
var digestUtil = esut.digestUtil;
var dc = require('mcp_db').dc;

var cons = require('mcp_constants');
var userType = cons.userType;
var msgStatus = cons.msgStatus;
var msgType = cons.msgType;

var config = require('mcp_config');
var ec = config.ec;
var prop = config.prop;
var ac = config.ac;

var service = require("mcp_service");
var digestSer = service.digestSer;
var moneySer = service.moneySer;
var termSer = service.termSer;

var AdminControl = function(){
    var self = this;
    self.cmd = {'AD01':1, 'AD02':2, 'AD03':3, 'AD04':4,
    'AD05':5, 'AD06':6, 'AD07':7, 'AD08':8, 'AD09':9, 'AD10':10};
    self.cmdArray = [{},
        {id:1, code:'AD01', fromType:prop.digestFromType.DB, des:"管理员登陆"},
        {id:2, code:'AD02', fromType:prop.digestFromType.CACHE, des:"获取权限树"},
        {id:3, code:'AD03', fromType:prop.digestFromType.CACHE, des:"修改消息"},
        {id:4, code:'AD04', fromType:prop.digestFromType.CACHE, des:"删除消息"},
        {id:5, code:'AD05', fromType:prop.digestFromType.CACHE, des:"修改期次"},
        {id:6, code:'AD06', fromType:prop.digestFromType.CACHE, des:"修改交易密钥"},
        {id:7, code:'AD07', fromType:prop.digestFromType.CACHE, des:"账户操作"},
        {id:8, code:'AD08', fromType:prop.digestFromType.CACHE, des:"新增用户"},
        {id:9, code:'AD09', fromType:prop.digestFromType.CACHE, des:"新增转发规则"},
        {id:10, code:'AD10', fromType:prop.digestFromType.CACHE, des:"期次开奖"}
    ];
};

AdminControl.prototype.handle = function(headNode, bodyStr, userCb)
{
    var self = this;
    async.waterfall([
        //是否是支持的cmd
        function(cb)
        {
            var cmd = self.cmd[headNode.cmd];
            if(cmd == undefined)
            {
                cb(ec.E9000);
            }
            else
            {
                cb(null);
            }
        },
        //校验头的用户类型是否合法
        function(cb)
        {
            var userTypeId = userType[headNode.userType];
            if(userTypeId == undefined)
            {
                cb(ec.E9005);
            }
            else
            {
                cb(null, userTypeId);
            }
        },
        //获得密钥
        function(userTypeId, cb)
        {
            var cmd = self.cmdArray[self.cmd[headNode.cmd]];
            digestSer.getKey({fromType:cmd.fromType, userId:headNode.userId, userType:userTypeId},
            function(err, key){
                cb(err, key);
            });
        },
        //先解密
        function(key, cb)
        {
            log.info(key);
            var decodedBodyStr = digestUtil.check(headNode, key, bodyStr);
            try {
                var bodyNode = JSON.parse(decodedBodyStr);
                headNode.key = key;
                cb(null, bodyNode);
            }
            catch (err)
            {
                cb(ec.E9003);
            }
        },
        //check the param
        function(bodyNode, cb){
            var method = 'check' + headNode.cmd;
            self[method](null, headNode, bodyNode, function(err){
                cb(err, bodyNode);
            });
        },
        //业务处理
        function(bodyNode, cb){
            var cmd = 'handle' + headNode.cmd;
            self[cmd](null, headNode, bodyNode, cb);
        }
    ], function (err, bodyNode) {
        userCb(err, bodyNode);
    });
};


AdminControl.prototype.checkAD01 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

AdminControl.prototype.checkAD02 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

AdminControl.prototype.checkAD03 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

AdminControl.prototype.checkAD04 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

AdminControl.prototype.checkAD05 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

AdminControl.prototype.checkAD06 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

AdminControl.prototype.checkAD07 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

AdminControl.prototype.checkAD08 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};


AdminControl.prototype.checkAD09 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};


AdminControl.prototype.checkAD10 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};



/**
 * find one's all operations
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
AdminControl.prototype.handleAD01 = function(user, headNode, bodyNode, cb)
{
    var stInfoTable = dc.mg.get("stInfo");
    stInfoTable.findOne({_id:headNode.userId}, {}, [], function(err, data){
        if(data)
        {
            var newSt = data.st;
            var now = new Date();
            if(now.getTime() - data.lastActiveTime > prop.loginExpiredSeconds*1000)
            {
                //expired
                newSt = digestUtil.createUUID();
            }
            stInfoTable.update({_id:headNode.userId}, {$set:{lastActiveTime:new Date().getTime(), st:newSt}},
            [], function(err, data){
                cb(null, {uniqueId:bodyNode.uniqueId, st:newSt});
            });
        }
        else
        {
            var st = digestUtil.createUUID();
            stInfoTable.save({_id:headNode.userId, st:st, lastActiveTime:new Date().getTime()}, [], function(err, data){
                cb(null, {uniqueId:bodyNode.uniqueId, st:st});
            });
        }
    });
};

/**
 * 获取权限树
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
AdminControl.prototype.handleAD02 = function(user, headNode, bodyNode, cb)
{
    var backBodyNode = {};
    var cond = bodyNode.cond;
    if(cond == undefined)
    {
        cond = {};
    }
    var userTypeId = userType[headNode.userType];
    cond.userType = userTypeId;
    var operationTable = dc.main.get("operation");
    operationTable.find(cond, {}).toArray(function(err, data){
        if(data)
        {
            backBodyNode.rst = data;
        }
        cb(null, backBodyNode);
    });
};

/**
 * 修改消息
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
AdminControl.prototype.handleAD03 = function(user, headNode, bodyNode, cb)
{
    var backBodyNode = {};
    var table = dc.mg_msg.get("msg");
    table.update(bodyNode.cond, bodyNode.data, [], function(err, data){
        if(err)
        {
            log.info(err);
            cb(ec.E0999);
        }
        else
        {
            cb(err, backBodyNode);
        }
    });
};

/**
 * 删除消息
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
AdminControl.prototype.handleAD04 = function(user, headNode, bodyNode, cb)
{
    var backBodyNode = {};
    var self = this;
    async.waterfall([
        function(cb)
        {
            var conn = dc.mg_msg.getConn();
            var table = conn.collection("msg");
            table.findAndRemove(bodyNode.cond, {}, [], function(err, data){
                if(err)
                {
                    cb(ec.E0999);
                }
                else
                {
                    cb(err, data);
                }
            });
        },
        //删除详细
        function(msg, cb)
        {
            if(msg.type == msgType.TERM)
            {
                var detail = dc.mg_msg.get("detail_term");
                detail.remove({msgId:msg._id}, [], function(err, data){
                    cb(err);
                });
            }
            else
            {
                cb(null);
            }
        }
    ], function (err, bodyNode) {
        cb(err, backBodyNode);
    });
};

/**
 * 修改期次
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
AdminControl.prototype.handleAD05 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    var table = dc.main.get("term");
    table.update(bodyNode.cond, bodyNode.data, [], function(err, data){
        if(err)
        {
            log.info(err);
            cb(ec.E0999);
        }
        else
        {
            cb(err, backBodyNode);
        }
    });
};

AdminControl.prototype.handleAD06 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    var table = dc.mg.get("stInfo");
    table.update(bodyNode.cond, bodyNode.data, [], function(err, data){
        if(err)
        {
            cb(ec.E0999);
        }
        else
        {
            log.info(data);
            cb(err, backBodyNode);
        }
    });
};

AdminControl.prototype.handleAD07 = function(user, headNode, bodyNode, cb)
{
    var backBodyNode = {};
    var pageSubjectId = bodyNode.subjectId;
    var subject = ac.getInfoById(pageSubjectId);
    var roleId = pageSubjectId.substr(0, 2);
    var accountId = pageSubjectId.substr(2, 2);
    var typeId = pageSubjectId.substr(4, 1);
    var subjectId = pageSubjectId.substr(5, 4);
    var amount = bodyNode.amount;
    var orderId = bodyNode.orderId;
    var entityId = bodyNode.entityId;
    moneySer.handleById(entityId, orderId, roleId, accountId, typeId, subjectId, amount, function(err, data){
        log.info(err);
        log.info(data);
        cb(err, backBodyNode);
    });
};


/**
 * 新增用户
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
AdminControl.prototype.handleAD08 = function(user, headNode, bodyNode, cb)
{
    var backBodyNode = {};
    async.waterfall([
        //新增用户
        function(cb)
        {
            var table = dc.main.get("customer");
            table.save(bodyNode, [], function(err, data){
                cb(err, bodyNode);
            });
        },
        //新增账户
        function(customer, cb)
        {
            var table = dc.mg.get("account");
            var account = {_id:customer.id, balance:0, version:0};
            table.save(account, [], function(err, data){
                cb(err, customer);
            });
        },
        //初始化登陆信息
        function(customer, cb)
        {
            var table = dc.mg.get("stInfo");
            var stInfo = {_id:customer.id, st:digestUtil.createUUID(),
                fixSt:digestUtil.createUUID(), version:0};
            table.save(stInfo, [], function(err, data){
                cb(err);
            });
        }
    ], function (err, bodyNode) {
        cb(err, backBodyNode);
    });
};

/**
 * 新增游戏转发规则
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
AdminControl.prototype.handleAD09 = function(user, headNode, bodyNode, cb)
{
    var backBodyNode = {};
    var data = bodyNode;
    data._id = data.customerId + "_" + data.gameCode;
    var table = dc.mg.get("relation");
    table.save(bodyNode, [], function(err, data){
        cb(err, backBodyNode);
    });
};


/**
 * 期次开奖
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
AdminControl.prototype.handleAD10 = function(user, headNode, bodyNode, cb)
{
    var backBodyNode = {};
    var table = dc.main.get("term");
    async.waterfall([
        //更新期次状态
        function(cb)
        {
            termSer.draw(bodyNode.termId, function(err, data){
                cb(err, data);
            });
        },
        //发送期次已经开奖的消息
        function(term, cb)
        {
            cb(null);
        }
    ], function (err, rst) {
        cb(err, backBodyNode);
    });
};


var adminControl = new AdminControl();
module.exports = adminControl;
