// mcp_control/PageControl.js
var dc = require('mcp_db').dc;
var esut = require('easy_util');
var digestUtil = esut.digestUtil;
var adminPageControl = require("./AdminPageControl.js");
var channelPageControl = require("./ChannelPageControl.js");
var termPageControl = require("./TermPageControl.js");
var gamePageControl = require("./GamePageControl.js");
var msgPageControl = require("./MsgPageControl.js");
var customerPageControl = require("./CustomerPageControl.js");
var sysPageControl = require("./SysPageControl.js");
var gameGradePageControl = require("./GameGradePageControl.js");

var PageControl = function(){};

PageControl.prototype.handle = function(headNode, bodyNode, cb)
{
    console.log(bodyNode);
    var self = this;
    var cmd = headNode.cmd;
    self[cmd[0]](headNode, bodyNode, cb);
};

PageControl.prototype.admin = function(headNode, bodyNode, cb)
{
    adminPageControl.handle(headNode, bodyNode, cb);
};


PageControl.prototype.game = function(headNode, bodyNode, cb)
{
    gamePageControl.handle(headNode, bodyNode, cb);
};


PageControl.prototype.term = function(headNode, bodyNode, cb)
{
    termPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.msg = function(headNode, bodyNode, cb)
{
    msgPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.customer = function(headNode, bodyNode, cb)
{
    customerPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.channel = function(headNode, bodyNode, cb)
{
    channelPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.sys = function(headNode, bodyNode, cb)
{
    sysPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.gamegrade = function(headNode, bodyNode, cb)
{
    gameGradePageControl.handle(headNode, bodyNode, cb);
};

var pageControl = new PageControl();
module.exports = pageControl;
